craft_parts/state_manager/step_state.py

```diff
@@ -76,15 +79,28 @@ class StepState(MigrationState, ABC):
 
     part_properties: dict[str, Any] = {}
     project_options: dict[str, Any] = {}
-
-    class Config:
-        """Pydantic model configuration."""
-
-        validate_assignment = True
-        extra = "ignore"
-        allow_mutation = False
-        alias_generator = lambda s: s.replace("_", "-")  # noqa: E731
-        allow_population_by_field_name = True
+    model_config = ConfigDict(
+        validate_assignment=True,
+        extra="ignore",
+        frozen=True,
+        alias_generator=lambda s: s.replace("_", "-"),
+        populate_by_name=True,
+    )
+
+    @model_validator(mode="after")
+    def _coerce_project_vars(self) -> Self:
+        """Coerce project_vars options to ProjectVar types."""
+        # FIXME: add proper type definition for project_options so that
+        # ProjectVar can be created by pydantic during model unmarshaling.
```

Comment:
Project var handling is another area that we'll need to refactor at some point.

Reply:
Agreed. As a scrivener's note, this (both the logic and the comment) came from `load_step_state` in `states.py` - a later PR removes that logic from there.